python-package/xgboost/spark/core.py

```diff
@@ -0,0 +1,810 @@
+import shutil
+import tempfile
+from typing import Iterator, Tuple
+import numpy as np
+import pandas as pd
+from scipy.special import expit, softmax
+from pyspark.ml import Estimator, Model
+from pyspark.ml.param.shared import HasFeaturesCol, HasLabelCol, HasWeightCol, \
+    HasPredictionCol, HasProbabilityCol, HasRawPredictionCol, HasValidationIndicatorCol
+from pyspark.ml.param import Param, Params, TypeConverters
+from pyspark.ml.util import MLReadable, MLWritable
+from pyspark.sql.functions import col, pandas_udf, countDistinct, struct
+from pyspark.sql.types import ArrayType, FloatType
+from xgboost import XGBClassifier, XGBRegressor
+from xgboost.core import Booster
+import cloudpickle
+import xgboost
+from xgboost.training import train as worker_train
+from .utils import get_logger, _get_max_num_concurrent_tasks
+from .data import prepare_predict_data, prepare_train_val_data, convert_partition_data_to_dmatrix
+from .model import (XgboostReader, XgboostWriter, XgboostModelReader,
+                    XgboostModelWriter, deserialize_xgb_model,
+                    get_xgb_model_creator, serialize_xgb_model)
+from .utils import (_get_default_params_from_func, get_class_name,
+                    HasArbitraryParamsDict, HasBaseMarginCol, RabitContext,
+                    _get_rabit_args, _get_args_from_message_list,
+                    _get_spark_session)
+
+from pyspark.ml.functions import array_to_vector, vector_to_array
+
+# Put pyspark specific params here, they won't be passed to XGBoost.
+# like `validationIndicatorCol`, `baseMarginCol`
+_pyspark_specific_params = [
+    'featuresCol', 'labelCol', 'weightCol', 'rawPredictionCol',
+    'predictionCol', 'probabilityCol', 'validationIndicatorCol'
+                                       'baseMarginCol'
+]
+
+_unsupported_xgb_params = [
+    'gpu_id',  # [ML-12862]
+]
+_unsupported_fit_params = {
+    'sample_weight',  # Supported by spark param weightCol
+    # Supported by spark param weightCol # and validationIndicatorCol
+    'eval_set',
+    'sample_weight_eval_set',
+    'base_margin'  # Supported by spark param baseMarginCol
+}
+_unsupported_predict_params = {
+    # [ML-12913], for classification, we can use rawPrediction as margin
+    'output_margin',
+    'validate_features',  # [ML-12923]
+    'base_margin'  # [ML-12689]
+}
+
+_created_params = {"num_workers", "use_gpu"}
+
+
+class _XgboostParams(HasFeaturesCol, HasLabelCol, HasWeightCol,
```

Comment:
Could we unify the parameter format? seems we have both formats like "featuresCol" and "max_bin"

Reply:
These params `HasFeaturesCol, HasLabelCol, HasWeightCol` comes from reusing pyspark ML params. These param name has been defined as "featureCol" etc. in pyspark ML param module. We cannot change the param name.

But, other params are coming from xgboost library (e.g. max_depth, max_bin), shall we make them to be camel-Case style format ?

Reply:
Yeah, IMO, we shouldn't change the parameters from xgboost library, since there is already a detailed description of the parameters, and better not to add another "same meaning" but "different format style" parameter.

``` python
class _XgboostParams(HasFeaturesCol, HasLabelCol, HasWeightCol,
                     HasPredictionCol, HasValidationIndicatorCol,
                     HasArbitraryParamsDict, HasBaseMarginCol):
```

I mean, could we re-define `HasFeaturesCol, HasLabelCol ...` in XGBoost pyspark wrapper instead of using the builit-in ones. I don't know if it makes sense.

Reply:
> could we re-define HasFeaturesCol, HasLabelCol ...

This might add a lot of code, I prefer to keep reusing HasFeaturesCol.., what about in xgboost estimator we link "feature_col"  param name to "featureCol"  ?

Reply:
Could we just leave HasXXXCol here and hack them in the constructor? eg,

users pass features_col and then convert to featuresCol internal which should be a black box? Does that make sense?